util: improve inspect performance

BridgeAR opened the pull request:
I reworked a couple of things in util inspect for performance. The code is fully covered.

__EDIT__:  Updated the comment here and benchmarks.

```                                                                         improvement confidence      p.value
 util/format.js type="no-replace" n=2000000                                1676.61 %        *** 6.038361e-60
 util/inspect-array.js type="denseArray_showHidden" len=100000 n=1000        87.89 %        *** 3.315801e-79
 util/inspect-array.js type="denseArray" len=100000 n=1000                   68.83 %        *** 5.648717e-52
 util/inspect-array.js type="mixedArray" len=100000 n=1000                   21.22 %        *** 2.438066e-52
 util/inspect-array.js type="sparseArray" len=100000 n=1000                   6.87 %        *** 9.551672e-50
 util/inspect-proxy.js n=1000000                                            418.34 %        *** 5.193464e-71
 util/inspect.js option="colors" method="Array" n=2000000                   121.92 %        *** 1.136217e-39
 util/inspect.js option="colors" method="Date" n=2000000                    294.42 %        *** 3.899375e-18
 util/inspect.js option="colors" method="Error" n=2000000                   507.31 %        *** 6.416333e-18
 util/inspect.js option="colors" method="Object_deep_ln" n=2000000          257.20 %        *** 8.804036e-23
 util/inspect.js option="colors" method="Object_empty" n=2000000            704.27 %        *** 8.422940e-46
 util/inspect.js option="colors" method="Object" n=2000000                  372.04 %        *** 3.430467e-19
 util/inspect.js option="colors" method="Set" n=2000000                     443.65 %        *** 7.503642e-30
 util/inspect.js option="colors" method="String_boxed" n=2000000            458.19 %        *** 4.199961e-23
 util/inspect.js option="colors" method="String_complex" n=2000000          875.46 %        *** 4.377511e-42
 util/inspect.js option="colors" method="String" n=5000000                 2209.03 %        *** 3.336942e-23
 util/inspect.js option="colors" method="TypedArray_extra" n=2000000        269.77 %        *** 6.353860e-22
 util/inspect.js option="colors" method="TypedArray" n=2000000              244.82 %        *** 1.157433e-47
 util/inspect.js option="none" method="Array" n=2000000                     298.39 %        *** 6.163207e-35
 util/inspect.js option="none" method="Date" n=2000000                      308.41 %        *** 3.828139e-26
 util/inspect.js option="none" method="Error" n=2000000                     499.46 %        *** 2.519987e-17
 util/inspect.js option="none" method="Object_deep_ln" n=2000000            470.60 %        *** 7.069515e-27
 util/inspect.js option="none" method="Object_empty" n=2000000              703.54 %        *** 5.821877e-41
 util/inspect.js option="none" method="Object" n=2000000                    678.48 %        *** 1.752143e-21
 util/inspect.js option="none" method="Set" n=2000000                       739.62 %        *** 5.402939e-18
 util/inspect.js option="none" method="String_boxed" n=2000000              483.94 %        *** 2.756513e-37
 util/inspect.js option="none" method="String_complex" n=2000000           1082.41 %        *** 3.900304e-64
 util/inspect.js option="none" method="String" n=5000000                   3954.65 %        *** 1.558337e-26
 util/inspect.js option="none" method="TypedArray_extra" n=2000000          410.59 %        *** 1.364578e-19
 util/inspect.js option="none" method="TypedArray" n=2000000                387.47 %        *** 1.789663e-23
 util/inspect.js option="showHidden" method="Array" n=2000000               272.92 %        *** 2.332172e-29
 util/inspect.js option="showHidden" method="Date" n=2000000                307.75 %        *** 4.581806e-16
 util/inspect.js option="showHidden" method="Error" n=2000000               254.46 %        *** 1.552997e-19
 util/inspect.js option="showHidden" method="Object_deep_ln" n=2000000      422.07 %        *** 4.184839e-46
 util/inspect.js option="showHidden" method="Object_empty" n=2000000        665.61 %        *** 2.095402e-24
 util/inspect.js option="showHidden" method="Object" n=2000000              626.02 %        *** 6.570478e-39
 util/inspect.js option="showHidden" method="Set" n=2000000                 815.83 %        *** 3.634105e-32
 util/inspect.js option="showHidden" method="String_boxed" n=2000000        680.82 %        *** 3.104074e-27
 util/inspect.js option="showHidden" method="String_complex" n=2000000     1058.45 %        *** 5.543800e-49
 util/inspect.js option="showHidden" method="String" n=5000000             3473.09 %        *** 6.496121e-29
 util/inspect.js option="showHidden" method="TypedArray_extra" n=2000000    345.93 %        *** 9.418444e-36
 util/inspect.js option="showHidden" method="TypedArray" n=2000000          316.96 %        *** 4.742310e-35
```

The main optimizations are
- Removed visibleKeys
- Removed proxy cache
- Removed Object.assign
- No key concatenating anymore
- No key recalculating anymore
- Improved indentation logic
- Improved string escape logic
- Added many fast paths
- Optimized code branches a lot
- Optimized (boxed) primitive handling
- Inline code if possible
- Only check extra keys if necessary

Besides that there were lots of smaller optimizations, the
code got a bit cleaned up and a few more tests got in.

Also two bug fixes came.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
util, benchmark

andrasq commented on lib/util.js:
```diff
@@ -361,168 +371,150 @@ function formatValue(ctx, value, recurseTimes) {
+    if (value.length === 0 && keyLength === 0)
+      return `${braces[0]}]`;
     formatter = formatArray;
+  } else if (isMapIterator(value)) {
```
a long series of if - else if - else if statements can also be formatted as a `switch (true) { ... }` (note that switch does a strict comparison, ie matches not truthy but `true`)

BridgeAR replied:
I you prefer that I can change it, otherwise I would just keep it as is.

addaleax commented on lib/util.js:
```diff
@@ -361,262 +371,229 @@ function formatValue(ctx, value, recurseTimes) {
+    if (ctx.seen === undefined) {
+      ctx.seen = [];
+      // A customInspect could be set only for partial objects.
+      // To prevent a breaking change, fill the lazy added array with all
```
grammar nit: lazy → lazily

BridgeAR replied:
:+1:

addaleax commented on lib/util.js:
```diff
@@ -361,262 +371,229 @@ function formatValue(ctx, value, recurseTimes) {
+  const customInspect = isCustomInspect(ctx, value);
+  if (customInspect !== undefined) {
+    if (ctx.seen === undefined) {
+      ctx.seen = [];
```
`ctx.seen = [...ctx.seenSet]`?

Also, I, personally, would be fine with having `seen` just be a getter/setter pair.

BridgeAR replied:
:+1:

BridgeAR replied:
Ref https://github.com/nodejs/node/pull/14881#discussion_r133975861

addaleax commented on lib/util.js:
```diff
@@ -300,12 +298,9 @@ function stylizeNoColor(str, styleType) {
 
 function arrayToHash(array) {
```
I feel like this could just be turned into a `Set` as well

BridgeAR replied:
I tried that and it was actually a bit slower in my benchmarks. When 6.1 lands I might have another look at it.

addaleax commented on lib/util.js:
```diff
@@ -318,6 +313,21 @@ function ensureDebugIsInitialized() {
 }
 
 
+function isCustomInspect(ctx, value) {
```
This really sounds like it would return a boolean ;)

BridgeAR replied:
:+1:

addaleax commented on lib/util.js:
```diff
@@ -361,262 +371,229 @@ function formatValue(ctx, value, recurseTimes) {
     }
   }
 
+  var type = typeof value;
```
const? Or maybe just keep it inlined, I think JS engines might be better at that?

BridgeAR replied:
Hm, I think it is nicer not to write `typeof x` many times over and `type` is currently redefined in [line 531](https://github.com/nodejs/node/pull/14881/files#diff-3deb3f32958bb937ae05c6f3e4abbdf5R531). But I can change it if you want me to.

TimothyGu replied:
Yeah V8 has explicit inlining for the form `typeof x === 'type'` -- but only for that form.

BridgeAR replied:
It is not about the performance. I think it is just nicer to read not having lots of `typeof` when this it is not necessary. But I guess you would like me to change that?

addaleax replied:
I mean, I don’t really have strong feelings about this, but yeah, I would say storing the result of `typeof` is kind of unidiomatic…

BridgeAR replied:
Fixed

addaleax commented on lib/util.js:
```diff
@@ -672,222 +644,209 @@ function formatError(value) {
+  for (i = 0; i < keys.length; i++) {
+    const key = keys[i];
+    // Skip positive numbers
+    if (+key >= 0 === false) {
```
Why not `!(+key >= 0)` or just `+key < 0`?

BridgeAR replied:
I like the short versions but I try to write more explicit JS and the latter does not work for keys other then numbers. So if you do not feel strongly about it, I would just keep it as is?

addaleax replied:
Your call, but, like, I do find `((some inequality expression) === false)` a bit odd.

BridgeAR replied:
I changed it to `!(+key >= 0)`

addaleax commented on lib/util.js:
```diff
@@ -672,222 +644,209 @@ function formatError(value) {
 
 
-function formatTypedArray(ctx, value, recurseTimes, visibleKeys, keys) {
+function formatTypedArray(ctx, value, recT, visibleKeys, keys) {
```
Wasn’t this a bit more readable before?

BridgeAR replied:
Addressed

addaleax commented:
Also, sorry for the merge conflict, I fixed Evan’s nit while landing my PR.

BridgeAR commented:
@addaleax I addressed your comments and don't worry about the conflicts, I was aware of those changes and I'll rebase after #14880 lands :)
Those are minor changes anyway.

BridgeAR commented on lib/util.js:
```diff
@@ -361,261 +371,202 @@ function formatValue(ctx, value, recurseTimes) {
-  }
+  // Using an array here is actually better for the average case then using
+  // a Set. `seen` will only check for the depth and will never grow to large.
+  if (ctx.seen.includes(value))
```
I reverted my change to a Set here: the `seen` part will only contain up to `ctx.depth + 1` elements and in the regular case that is three. Using a Set for such a low amount of elements seems unnecessary, as it increases the code complexity due to having to handle both, an array and a set in case customInspect is set.

BridgeAR commented:
I rebased and added a new array fast path. So PTAL

Just for fun - this is how the proofiling looks like with `Array(1e5).fill(1)`

```
 [Summary]:
   ticks  total  nonlib   name
    261    0.6%    0.6%  JavaScript
  39502   94.0%   94.3%  C++
   1797    4.3%    4.3%  GC
    129    0.3%          Shared libraries
   2147    5.1%          Unaccounted

 [C++ entry points]:
   ticks    cpp   total   name
  37497   99.2%   89.2%  v8::internal::Runtime_ObjectKeys(int, v8::internal::Object**, v8::internal::Isolate*)
     75    0.2%    0.2%  v8::internal::Builtin_ObjectGetOwnPropertyDescriptor(int, v8::internal::Object**, v8::internal::Isolate*)
```

0.6 % JavaScript :rofl:

BridgeAR commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9767/

BridgeAR commented:
So I pushed a bit more code and updated the benchmarks. The result is pretty decent now.

BridgeAR commented on lib/util.js:
```diff
@@ -211,28 +211,28 @@ function debuglog(set) {
  */
-/* legacy: obj, showHidden, depth, colors*/
-function inspect(obj, opts) {
+function inspect(obj, opts, /* legacy */depth, /* legacy */colors) {
```
Well, I do not have a strong opinion.
I went ahead and changed it back to a similar way as before to make you happy ;-)

BridgeAR commented:
Ok, the new benchmarks are outdated again as I was able to improve some more things that had significant impact on `showHidden`. So just think of most showHidden results +100%.

BridgeAR commented:
So after looking at it again I added some final changes. The benchmarks are updated in the main part. As the code still changed significantly, please take a close look again. @addaleax @refack @andrasq 

And I have an issue with one test running this locally but I can not explain why the test is failing. I get the following output. @bnoordhuis you wrote that test, would you be so kind and have a look at it? Because I do not see the connection of the error to my changes...

```
	async-hooks doctool inspector known_issues message parallel pseudo-tty sequential \
	addons addons-napi
[02:05|%  91|+ 1642|-   0]: release stack_overflow_async length differs.       
expect=4
actual=2
patterns:
pattern = ^RangeError\:\ Maximum\ call\ stack\ size\ exceeded$
pattern = ^\ \ \ \ at\ f\ \(.*test.*message.*stack\_overflow\_async\.js\:.*\)$
pattern = ^\ \ \ \ at\ f\ \(.*test.*message.*stack\_overflow\_async\.js\:7\:.*\)$
pattern = ^\ \ \ \ at\ f\ \(.*test.*message.*stack\_overflow\_async\.js\:7\:.*\)$
outlines:
outline = (node:21361) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): RangeError: Maximum call stack size exceeded
outline = (node:21361) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
=== release stack_overflow_async ===                    
Path: message/stack_overflow_async
(node:21361) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): RangeError: Maximum call stack size exceeded
(node:21361) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Command: out/Release/node --stack_trace_limit=3 /home/ruben/repos/node/node/test/message/stack_overflow_async.js
[02:46|% 100|+ 1785|-   1]: Done                                               
```

addaleax commented:
@BridgeAR That’s a really new test, and I’m pretty sure it’s not related to your PR

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/9817

BridgeAR commented:
@addaleax the weird part is that is it happening only with this PR. On master it does not fail. It is reproducible for me on my local machine and also the CI fails exactly at this test.

BridgeAR commented:
@bnoordhuis it would be nice if you could have a look at this PR because it somehow breaks [this test](https://github.com/nodejs/node/blob/master/test/message/stack_overflow_async.js) even though there should not be any connection. The only way for me to fix the test is by calling `process.stdout.write("")` (or `console.log()`) somewhere before the test starts.

BridgeAR commented:
@nodejs/collaborators PTAL

benjamingr commented on lib/internal/util.js:
```diff
@@ -257,6 +257,18 @@ function promisify(original) {
 promisify.custom = kCustomPromisifiedSymbol;
 
+// The build in join is relatively slow (currently v8 6.0)
+function join(output, separator) {
```
This is weird, any idea why it's slower?

BridgeAR replied:
That is a known issue in v8. There is a open ticket - I am a bit in a hurry but I can add the reference to it later on.

benjamingr replied:
@BridgeAR - no rush - I'm just curious - I approved the actual code changes - but I'd like to make sure this passes CITGM before release and I'm afraid that merging it would mean that it might be missed (since this is semver patch).

BridgeAR replied:
https://bugs.chromium.org/p/v8/issues/detail?id=2897

It does not seem to be on the priority list though :smile:

mscdex commented on benchmark/util/inspect.js:
```diff
@@ -3,14 +3,96 @@ var util = require('util');
+    case 'Object_deep_ln':
+      if (options)
+        options.depth = Infinity;
+      obj = { frist:
```
typo

mscdex commented on lib/internal/util.js:
```diff
@@ -257,6 +257,18 @@ function promisify(original) {
 
 promisify.custom = kCustomPromisifiedSymbol;
 
+// The build in join is relatively slow (currently v8 6.0)
```
I would replace this with something like: 'array.join in v8 6.0 is slower'

Also, adding a link to the issue in the comment would be beneficial.

ronkorving replied:
Also, typo? [built-in](https://en.wiktionary.org/wiki/built-in)

tniessen commented on lib/util.js:
```diff
@@ -383,511 +428,437 @@ function formatValue(ctx, value, recurseTimes) {
-  if (empty === true) {
-    return `${braces[0]}${base}${braces[1]}`;
-  }
+  // Using an array here is actually better for the average case then using
```
Typo: `than`

tniessen commented on lib/internal/util.js:
```diff
@@ -257,6 +257,18 @@ function promisify(original) {
 
+// The build in join is relatively slow (currently v8 6.0)
+function join(output, separator) {
+  var str = '';
```
Any specific reason to use `var` instead of `let` in the whole PR?

BridgeAR replied:
I did not measure `let` and `var` recently but in general using let was slower then var. I have to check at some point if that is still the case. I would rather not change it right to `let` in this PR though.

tniessen commented on lib/util.js:
```diff
@@ -383,511 +428,437 @@ function formatValue(ctx, value, recurseTimes) {
+      braces[0] = `${ctorName}{`;
+      formatter = formatPromise;
+    } else {
+      // Check boxed primitives other then string with valueOf()
```
typo: `other than`

tniessen commented on lib/util.js:
```diff
@@ -383,511 +428,437 @@ function formatValue(ctx, value, recurseTimes) {
+      output.push(`[${extraKeys[i]}]: ${str}`);
     }
   }
+  // TypedArrays can not have holes. Therefore it is safe to assume that all
```
typo: `cannot`

mscdex commented:
@BridgeAR How much of a difference does the custom `join()` make here? I wonder if there is something that could be done upstream in V8? /cc @bmeurer ?

TimothyGu commented:
Re. `join()`: I'd rather not include a special optimized version of it in our code base. `util.inspect()` has never been fast, so as long as there's no performance regression (which seems unlikely given how slow everything is right now in the benchmark) I think we should value readability over the last drop of performance at this moment.

bmeurer commented:
`Array#join` is definitely a know performance pain point, and we will work on it at some point, just unclear how quickly that's going to happen. In other words: Don't hold your breath for it.

BridgeAR commented:
@TimothyGu as @bmeurer pointed out, it might take a while to get this in. Otherwise I am fully with you. Would you still want me to remove it?

@mscdex I did not check how much difference the individual change did for the benchmarks as that would have required running them for a while. In general `Array#join` is about 40% slower than the function that I added.

BridgeAR commented:
Comments addressed

TimothyGu commented:
@BridgeAR That was certainly not some great news from @bmeurer, but my stance on it remains the same, if that is alright with you

BridgeAR commented:
@TimothyGu I just benchmarked a regular array and it is 6% faster to use the individual function. Since arrays, set, map and objects with many properties are the "slow" parts in general, I think it would be nice to keep the function around. I could add a TODO so someone will definitely stumble open it again instead?

```
// node benchmark/compare.js --new ./node-join --old out/Release/node --filter array --set type=denseArray --set len=100 --set n=5e5 util
util/inspect-array.js type="denseArray" len=100 n=500000      6.17 %        *** 2.281457e-06
```

BridgeAR commented:
I checked some weird edge cases and found two regressions that I fixed plus a bug that we were never aware of. I fixed all those and updated the code accordingly including some minor cleanup.

TimothyGu commented on test/parallel/test-util-inspect.js:
```diff
@@ -327,6 +317,61 @@ assert.strictEqual(
+// Indices out of bounds
+{
+  const arr = [];
+  arr[Math.pow(2, 32)] = true; // not a valid array index
```
`2 ** 32`?

TimothyGu commented on lib/util.js:
```diff
@@ -282,612 +364,515 @@ inspect.styles = Object.assign(Object.create(null), {
+        break;
+      const index = +key;
+      // Arrays can only have up to 2^32 - 1 entries
+      if (index > 4294967294)
```
`2 ** 32 - 2`? V8 should be able to constant-fold?

BridgeAR commented:
Comments addressed

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,78 @@ const errorToString = Error.prototype.toString;
+};
+
+// Escape control characters and the backslash.
+// This is similar to JSON stringify escaping.
```
You should at least mention the V8 version in here because the magic values used within could be tied to specific versions.

BridgeAR replied:
I am not sure what magic values you mean. We define what characters we want to escape.

mscdex replied:
Immediately right at the top of the function there is a comment that reads:

```js
// Some magic numbers that worked out fine while benchmarking
```

BridgeAR replied:
Right, I forgot my comment

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,78 @@ const errorToString = Error.prototype.toString;
+
+// Table of character substitutions for text
+const escapeCodesText = {
+  39: "\\'",
```
I wonder if using a sufficiently wide array (at least 128 to cover all of ASCII) would improve performance compared to a plain object like this. If not, it might be good to also compare against an object created with `Object.create(null)`.

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,78 @@ const errorToString = Error.prototype.toString;
+  12: '\\f'
+};
+// Table of character substitutions for keys
+const escapeCodesKeys = {
```
Ditto.

BridgeAR commented:
@mscdex I changed the meta object to an array. It is a tad faster than the object even though there is not much difference overall. But I liked that the function became smaller by doing so.

I would like to land this soon, so please all take another look. I think there is little left to be changed right now.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10049/

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+    return `'${str}'`;
+  if (str.length > 50) {
+    return `'${str.replace(escapeSequencesRegExp, (str) => {
+      return meta[str.charCodeAt(0)];
```
This won't work for non-ASCII characters. You will need to do `meta[str.charCodeAt(0)]` first, and check if it's `undefined` or `''`. If it is one of those two, just return an empty string.

BridgeAR replied:
I am only looking for characters to escape. Therefore the replace function will only be triggered for wanted characters.

BridgeAR replied:
As a side note - checking for `undefined` would be a bad idea because we would have a out of bounds access and that is bad for the performance.

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+  if (str.length < 5000 && textRegExp.test(str))
+    return `'${str}'`;
+  if (str.length > 50) {
+    return `'${str.replace(escapeSequencesRegExp, (str) => {
```
Using a plain loop (with a `charCodeAt()` range check) instead *might* be faster than `replace()` + regexp?

BridgeAR replied:
That part is explicitly added for strings with a length above 50 because of my benchmarks. It highly depended on the used input but in general the replace function wins for bigger string length.

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+// eslint-disable-next-line
+const escapeSequencesRegExp = /[\\'\x00-\x1f]/g;
+const keyStrRegExp = /^[a-zA-Z_][a-zA-Z_0-9]*$/;
+const textRegExp = /^[a-zA-Z_0-9 ,.?!+\-&"öäüÜÖÄß():;<>{}[\]=$]*$/;
```
Where did this regexp come from?

BridgeAR replied:
I came up with it. I tried to think of any "somewhat common" characters that do not have to be escaped but might be used anyway. I benchmarked the result performance and it does not really change from using `/^[a-zA-Z_0-9 ,.?!+\-&"öäüÜÖÄß():;<>{}[\]=$]*$/` to e.g. `/^[a-zA-Z_0-9 ,.?!+\-$]*$/`

mscdex replied:
All unicode characters can be inserted unescaped for example, that's why I was confused about the inclusion of only a handful of them. www.json.org shows the characters allowed in a string literal. I think the list of characters can be shortened/simplified.

BridgeAR replied:
Originally I used `escapeSequencesRegExp` but it turned out to be slower than using `textRegExp`. As it is somewhat about common cases I used this one. I also tried to point out that performance wise it does not make any difference if the RegExp is further simplified.

mscdex replied:
FWIW I just ran some quick benchmarks with simple inputs and using a simpler, negated regexp check (e.g. `/[\x00-\x1F\x7F']/`) seems to be noticeably quicker for strings containing a character needing to be escaped. For strings that don't need escaping, a simple loop over the string was measurably faster, but the performance difference there is not as large as the negated regexp in the other case.

BridgeAR replied:
Hm, yes. Somehow I messed up my benchmark about only checking the escape sequences. I only added that extra RegExp because I was certain my benchmark results were better for the individual one but when benchmarking it again I confirm your results.

However, I do remember why I used a RegExp in general and not use the loop as default: the performance is somewhat consistent with the RegExp but with the loop it becomes more and more expensive in case there are characters that have to be escaped. I will check if I find a alternative way way to get the common case even better.

BridgeAR replied:
I realized that I made a mistake while originally benchmarking the RegExp and therefore I had the impression the individual one would be faster.

I also had another look at the loop against replace and I increased the switch to 100 chars. It seems to be the better break even point.

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/981/

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+  if (str.length < 5000 && textRegExp.test(str))
+    return `'${str}'`;
+  if (str.length > 50) {
+    return `'${str.replace(escapeSequencesRegExp, (str) => {
```
We could extract this callback to avoid any potential closures created by V8 for each call to this function.

BridgeAR replied:
:+1:

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+  for (var i = 0; i < str.length; i++) {
+    const point = str.charCodeAt(i);
+    if (point === 39 || point === 92 || point < 32) {
+      result += `${str.slice(last, i)}${meta[point]}`;
```
You might consider adding a check that `last < i` to avoid unnecessarily slicing to an empty string. That's typically what I've done in other parts of core.

BridgeAR replied:
I am not a big fan of adding this extra check as it makes the code more complicated for rare cases but I will add the check nevertheless.

mscdex replied:
I wouldn't think they would really be that rare. Consecutive characters like tabs or newlines for example could be more common. Adding the extra check may/may not be worth it though, it would need some benchmarking to confirm my suggestion improves things.

BridgeAR replied:
Changed.

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+  if (last === 0) {
+    result = str;
+  } else {
+    result += str.slice(last);
```
Similarly here, make sure that `last` is not `str.length` to avoid unnecessary slicing.

BridgeAR replied:
As the comment above.

mscdex commented on lib/util.js:
```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
+}
+
+// Escape control characters and single quotes.
+function keyEscape(str) {
```
Ditto about checking `last` and `i` in this function.

BridgeAR replied:
See above.

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
-    return ctx.stylize(`'${simple}'`, 'string');
-  }
+function formatPrimitive(fn, value) {
+  const type = typeof value;
```
You might double check the performance with this. I know this would cause a performance hit when saving a `typeof` to a variable and checking that variable because V8 used to (and maybe still) would *inline* such checks when doing `typeof foo === 'string'`, but it cannot do that when it is saved to a variable.

BridgeAR replied:
I checked and my results showed exactly no difference for one way or the other. I have the feeling quite some people prefer always using typeof though, so I went ahead and changed it.

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
-      output.push(ctx.stylize(message, 'undefined'));
-      index = i;
-      if (++visibleLength === maxLength)
+// The array is sparse and or has extra keys
```
Minor nit: should be 'and/or' instead of 'and or'

BridgeAR replied:
:+1:

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
+      if (index > 2 ** 32 - 2)
         break;
+      if (i !== index) {
+        if (!numberRegExp.test(key))
```
I wonder if it's worthwhile avoiding this regexp for the first key since it is already checked earlier in the function: `numberRegExp.test(keys[0])`

mscdex replied:
Also, perhaps we can avoid using the regexp since we already check the numerical value against an upper limit. Checking for `NaN` should be easy with something like `index !== index`.

BridgeAR replied:
That is a very unlikely case - it will only be possible for sparsed arrays that are sparsed in the beginning. I am against adding that check.

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
+    var key;
+    for (i = keys.length - 1; i >= 0; i--) {
+      key = keys[i];
+      if (numberRegExp.test(key) && +key < 2 ** 32 - 1)
```
Ditto about removing use of regexp here and simply checking for `NaN`.

BridgeAR replied:
Using NaN is not sufficient here, we have to use the RegExp as e.g. `+004 === 4` and that is wrong.

BridgeAR replied:
Oh, and if I remember correct it was even a tad faster to use the RegExp.

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
+    return formatSpecialArray(ctx, value, recurseTimes, keys, len, valLen);
 
-function formatTypedArray(ctx, value, recurseTimes, visibleKeys, keys) {
+  const output = [];
```
Maybe this could be changed to a "fixed length" Array, since we can determine up-front how many elements there will be (based on `remaining` and `ctx.showHidden`)?

BridgeAR replied:
I still want to benchmark this at some point but I am actually thinking about doing the opposite in general due to the newest v8 blog post (using a predefined Array size results in the internal holey type and that type will not change even when the array became dense).

mscdex replied:
I thought that was fixed in a previous version of V8 already? /cc @bmeurer

BridgeAR replied:
Interesting. I benchmarked it and the results are better with `new Array(len)` as expected. The [new blogpost](https://v8project.blogspot.com.br/2017/09/elements-kinds-in-v8.html) says the opposite though...

Anyway - I changed it to `new Array` in almost all places.

mscdex commented on lib/util.js:
```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
-    output.push(str);
-  });
+function formatSet(ctx, value, recurseTimes, keys) {
+  const output = [];
```
Same here, this could be made a "fixed array."

BridgeAR replied:
See above.

mscdex commented on lib/util.js:
```diff
@@ -71,10 +71,13 @@ const errorToString = Error.prototype.toString;
+const strEscapeSequencesRegExp = /[\x00-\x1f\x27\x5c]/;
+const keyEscapeSequencesRegExp = /[\x00-\x1f\x27]/;
+const strEscapeSequencesReplacer = /[\x00-\x1f\x27\x5c]/g;
+const keyEscapeSequencesReplacer = /[\x00-\x1f\x27]/g;
```
Is there really a performance difference between the 'g' modifier and without with `.test()`? If not, we shouldn't duplicate these regexps.

BridgeAR replied:
const re = /test/g

console.log(re.test("tester")) // true
console.log(re.test("tester")) // false

mscdex commented on lib/util.js:
```diff
@@ -95,50 +98,59 @@ const meta = [
   if (last === 0) {
-    result = str;
-  } else {
+    result += str;
```
Why the change to concatenate with an empty string? I think the assignment is a better choice.

BridgeAR replied:
There was no reason for it - I was playing around. I changed it back

BridgeAR commented:
One more run https://ci.nodejs.org/job/node-test-pull-request/10064/

BridgeAR commented:
Any further remarks from anyone? Otherwise I will land this tomorrow.

BridgeAR commented:
I rebased this. It would be nice if someone could just give a new LG before I land this as it was a couple of days and some code still changed since then @TimothyGu @mcollina @mscdex @addaleax @benjamingr @refack.

I get 504 errors for the CI but the tests are passing locally and each commit is passing on its own. As this is not platform specific code we do not have to rely on the CI in this case.

mscdex commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10086/

BridgeAR commented:
The linter had a hiccup and I restarted that job https://ci.nodejs.org/job/node-test-linter/11783/ :heavy_check_mark:

BridgeAR commented:
Landed in 6bfc439711daf4361f7ce6d9256739e1a2c07141 ... 8f52ccc828c26aee8fe78c82a8a23fabd21918b7. Thanks for all the reviews :smiley:

jasnell commented:
This needs a backport pr to land in v8.x

MylesBorins commented:
should this land in LTS? If so it will need to bake a bit longer.

Please change labels as appropriate
